wapitiCore/attack/mod_drupal_enum.py

```diff
@@ -0,0 +1,117 @@
+import re
+import json
+import hashlib
+from os.path import join as path_join
+
+from wapitiCore.net.web import Request
+from wapitiCore.attack.attack import Attack
+from wapitiCore.language.vulnerability import LOW_LEVEL, _
+from wapitiCore.definitions.fingerprint import NAME as TECHNO_DETECTED
+
+MSG_TECHNO_VERSIONED = _("{0} {1} detected")
+MSG_NO_DRUPAL = _("No Drupal Detected")
+
+class mod_drupal_enum(Attack):
+    """Detect Drupal version."""
+    name = "drupal_enum"
+    PAYLOADS_HASH = "drupal_hash_files.json"
+    PAYLOADS_FILE_THEMES = "wordpress_themes.txt"
+
+    versions = []
+    def __init__(self, crawler, persister, logger, attack_options):
+        Attack.__init__(self, crawler, persister, logger, attack_options)
+        self.finished = False
+
+    def get_hash(self):
+        with open(path_join(self.DATA_DIR, self.PAYLOADS_HASH), errors="ignore") as hashes:
+            data = json.load(hashes)
+            return data
+
+    def detect_version(self, url):
+        vers = {}
+        data = self.get_hash()
+        for uri in data :
+            req = Request('{}{}'.format(url, uri))
+            rep = self.crawler.get(req)
+            if rep.status != 200:
+                continue
+
+            cont = rep.content.encode()
+            hash_content = hashlib.sha256(cont).hexdigest()
+
+            if uri in data :
+                if hash_content in data[uri] :
+                    vers[uri] = data[uri][hash_content]
+
+        i = 0
+        result =[]
+        for key in vers :
+            if i == 0 :
+                result = list(vers[key])
+                i = i+1
+            else :
+                result = list(set(result) & set(vers[key]))
+        self.versions = self.versions + result
+
+    def check_drupal(self, url):
+        check_list = ['sites/', 'core/misc/drupal.js', 'misc/drupal.js', 'misc/test/error/404/ispresent.html']
+        for item in check_list :
+            req = Request('{}{}'.format(url, item))
+            rep = self.crawler.get(req)
+            if rep.status != 404:
+                return True
+        return False
+
+
+    def must_attack(self, request: Request):
+        if self.finished:
+            return False
+
+        if request.method == "POST":
+            return False
+        return request.url == self.persister.get_root_url()
+
+    def attack(self, request: Request):
+        self.finished = True
+        request_to_root = Request(request.url)
+        response = self.crawler.send(request_to_root, follow_redirects=True)
+
+        if self.check_drupal(request_to_root.url):
+            self.detect_version(request_to_root.url)
+            if self.versions :
+                self.versions = sorted(self.versions, key=lambda x: [i for i in x.split('.')])
```

Comment:
You may be interested in the Python `natsort` module but if should be enough here

Reply:
I tried to do it with natsort  but i had an import problem